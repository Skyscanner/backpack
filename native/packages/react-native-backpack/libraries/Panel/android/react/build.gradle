apply plugin: 'com.android.library'
apply plugin: 'kotlin-android'

android {
  compileSdkVersion rootProject.ext.compileSdkVersion
  buildToolsVersion rootProject.ext.buildToolsVersion


  defaultConfig {
    minSdkVersion rootProject.ext.minSdkVersion
    targetSdkVersion rootProject.ext.targetSdkVersion
    versionCode 1
    versionName "1.0"
  }
}

dependencies{
  implementation "com.facebook.react:react-native:+" // From node_modules
  implementation project(':panel-native')
  implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
}
repositories {
  mavenCentral()
}
