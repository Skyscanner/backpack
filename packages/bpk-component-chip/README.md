# bpk-component-chip

> Backpack chip component.

## Installation

Check the main [Readme](https://github.com/skyscanner/backpack#usage) for a complete installation guide.

## Usage

```js
import React from 'react';
import BpkSelectableChip, { BpkDismissibleChip, BpkDropdownChip, CHIP_TYPES } from '@skyscanner/backpack-web/bpk-component-chip';
import BeachIconSm from '@skyscanner/backpack-web/bpk-component-icon/sm/beach';

export default () => (

  <div style={{ display: 'flex' }}> // IMPORTANT: Flex styles make sure chips align with each other
    // Standard selectable chip.
    <BpkSelectableChip
      accessibilityLabel="Press to toggle chip"
      selected={false}
      onClick={() => { /* Use state to set 'selected={true}' */ }}
    >
      Toggle me
    </BpkSelectableChip>

    // Selectable chip with an icon.
    <BpkSelectableChip
      accessibilityLabel="Press to toggle chip"
      selected={false}
      onClick={() => { /* Use state to set 'selected={true}' */ }}
      leadingAccessoryView={<BeachIconSm />}
    >
      Toggle me
    </BpkSelectableChip>

    // Standard dropdown chip.
    <BpkDropdownChip
      accessibilityLabel="Press to toggle chip"
      selected={false}
      onClick={() => { /* Use state to set 'selected={true}' */ }}
    >
      Toggle me
    </BpkDropdownChip>

    // Dropdown chip with an icon.
    <BpkDropdownChip
      accessibilityLabel="Press to toggle chip"
      selected={false}
      onClick={() => { /* Use state to set 'selected={true}' */ }}
      leadingAccessoryView={<BeachIconSm />}
    >
      Toggle me
    </BpkDropdownChip>

    // Standard dismissible chip.
    <BpkDismissibleChip
      accessibilityLabel="Press to dismiss chip"
      onClick={() => { /* Use state to handle removing this chip. */ }}
    >
      Dismiss me
    </BpkDismissibleChip>

    // Dismissible chip with an icon.
      <BpkDismissibleChip
      accessibilityLabel="Press to dismiss chip"
      onClick={() => { /* Use state to handle removing this chip. */ }}
      leadingAccessoryView={<BeachIconSm />}
    >
      Dismiss me
    </BpkDismissibleChip>
  </div>
);
```

## Props

### BpkSelectableChip

| Property              | PropType                                                              | Required | Default Value |
| --------------------- | --------------------------------------------------------------------- | -------- | ------------- |
| accessibilityLabel    | string                                                                | true     | -             |
| children              | node                                                                  | true     | -             |
| onClick               | func                                                                  | true     | -             |
| className             | string                                                                | false    | null          |
| disabled              | bool                                                                  | false    | false         |
| leadingAccessoryView  | node                                                                  | false    | null          |
| selected              | bool                                                                  | false    | false         |
| trailingAccessoryView | node                                                                  | false    | null          |
| type                  | oneOf(`CHIP_TYPES.onDark`, `CHIP_TYPES.default`, `CHIP_TYPES.onImage`) |

### BpkDropdownChip

| Property              | PropType                                                              | Required | Default Value |
| --------------------- | --------------------------------------------------------------------- | -------- | ------------- |
| accessibilityLabel    | string                                                                | true     | -             |
| children              | node                                                                  | true     | -             |
| onClick               | func                                                                  | true     | -             |
| className             | string                                                                | false    | null          |
| disabled              | bool                                                                  | false    | false         |
| leadingAccessoryView  | node                                                                  | false    | null          |
| selected              | bool                                                                  | false    | false         |
| type                  | oneOf(`CHIP_TYPES.onDark`, `CHIP_TYPES.default`, `CHIP_TYPES.onImage`) |

### BpkDismissibleChip

Dismissible chips are selectable chips that have been preconfigured to have a 'close' icon trailing accessory view and cannot be selected, so they have the same props as `BpkSelectableChip`, minus `trailingAccessoryView` and `selected`.

| Property             | PropType                                                              | Required | Default Value |
| -------------------- | --------------------------------------------------------------------- | -------- | ------------- |
| accessibilityLabel   | string                                                                | true     | -             |
| children             | node                                                                  | true     | -             |
| onClick              | func                                                                  | true     | -             |
| className            | string                                                                | false    | null          |
| leadingAccessoryView | node                                                                  | false    | null          |
| type                 | oneOf(`CHIP_TYPES.onDark`, `CHIP_TYPES.default`, `CHIP_TYPES.onImage`) |

## Theme Props

### Default

- `chipDefaultSelectedBackgroundColor`
- `chipDefaultSelectedTextColor`

### On Dark

- `chipOnDarkSelectedBackgroundColor`
- `chipOnDarkSelectedTextColor`

### On Image

- `chipOnImageSelectedBackgroundColor`
- `chipOnImageSelectedHoverBackgroundColor`
- `chipOnImageSelectedActiveBackgroundColor`
- `chipOnImageSelectedTextColor`
