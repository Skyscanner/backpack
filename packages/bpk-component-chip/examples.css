@keyframes bpk-keyframe-spin{100%{transform:rotate(1turn)}}.bpk-chip-examples__wrapper{display:flex;flex-wrap:wrap}.bpk-chip-examples__chip{margin-right:.25rem;margin-bottom:.25rem}
