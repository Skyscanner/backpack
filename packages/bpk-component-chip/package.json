{
  "name": "bpk-component-chip",
  "version": "2.0.80",
  "description": "Backpack chip component.",
  "main": "index.js",
  "repository": {
    "type": "git",
    "url": "git@github.com:Skyscanner/backpack.git"
  },
  "author": "Backpack Design System <backpack@skyscanner.net>",
  "license": "Apache-2.0",
  "peerDependencies": {
    "react": "^15.0.0 || ^16.0.0"
  },
  "dependencies": {
    "bpk-component-close-button": "^1.0.127",
    "bpk-mixins": "^17.21.0",
    "bpk-react-utils": "^2.7.6",
    "prop-types": "^15.6.2"
  },
  "gitHead": "edd14f16dc571cc009d908c3c8f7d27a428b6de8",
  "sideEffects": false
}
