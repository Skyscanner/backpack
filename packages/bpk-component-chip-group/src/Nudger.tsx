/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { type MutableRefObject, useEffect, useState } from 'react';

import { BpkButtonV2, BUTTON_TYPES } from '../../bpk-component-button';
import { CHIP_TYPES } from '../../bpk-component-chip';
import { withButtonAlignment } from '../../bpk-component-icon/index';
import ArrowLeft from '../../bpk-component-icon/sm/long-arrow-left';
import ArrowRight from '../../bpk-component-icon/sm/long-arrow-right';
import { cssModules, isRTL } from '../../bpk-react-utils/index';

import type { ChipStyleType } from './BpkMultiSelectChipGroup';

import STYLES from './Nudger.module.scss';

const getClassName = cssModules(STYLES);


const CHIP_STYLE_TO_BUTTON_STYLE = {
  [CHIP_TYPES.default]: BUTTON_TYPES.secondary,
  [CHIP_TYPES.onDark]: BUTTON_TYPES.secondaryOnDark,
  [CHIP_TYPES.onImage]: BUTTON_TYPES.primaryOnDark,
}

export const POSITION = {
  leading: 'leading',
  trailing: 'trailing',
} as const;

type Props = {
  ariaLabel: string;
  chipStyle?: ChipStyleType;
  scrollContainerRef: MutableRefObject<HTMLElement | null>;
  position: (typeof POSITION)[keyof typeof POSITION];
}

const AlignedLeftArrowIcon = withButtonAlignment(ArrowLeft);
const AlignedRightArrowIcon = withButtonAlignment(ArrowRight);

// Chosen based on feeling good with the example stories
const SCROLL_DISTANCE = 150;

const Nudger = ({
  ariaLabel,
  chipStyle = CHIP_TYPES.default,
  position,
  scrollContainerRef
}: Props) => {
  const [show, setShow] = useState(false);
  const [enabled, setEnabled] = useState(false);

  const leading = position === POSITION.leading;
  const rtl = isRTL();
  const isLeft = (leading && !rtl) || (!leading && rtl);

  useEffect(() => {
    const interval = setInterval(() => {
      if (!scrollContainerRef.current) {
        return;
      }

      const { offsetWidth, scrollLeft, scrollWidth } = scrollContainerRef.current;
      const scrollValue = rtl ? -Math.floor(scrollLeft) : Math.ceil(scrollLeft);
      const showLeading = scrollValue > 0;
      const showTrailing = scrollValue < scrollWidth - offsetWidth;

      setShow(showLeading || showTrailing);
      setEnabled((leading && showLeading) || (!leading && showTrailing))
    }, 100);
    return () => clearInterval(interval);
  }, [leading, rtl, scrollContainerRef]);

  const classNames = getClassName(
    'bpk-chip-group-nudger',
    `bpk-chip-group-nudger--${leading ? "leading" : "trailing"}`,
  )

  return show ? (
    <div className={classNames}>
      <BpkButtonV2
        title={ariaLabel}
        type={CHIP_STYLE_TO_BUTTON_STYLE[chipStyle]}
        iconOnly
        disabled={!enabled}
        onClick={() => {
          if (scrollContainerRef.current) {
            scrollContainerRef.current.scrollBy({
              left: isLeft ? -SCROLL_DISTANCE : SCROLL_DISTANCE,
              behavior: 'smooth',
            });
          }
        }}
      >
        {isLeft ? <AlignedLeftArrowIcon /> : <AlignedRightArrowIcon />}
      </BpkButtonV2>
    </div>
  ) : null;
}

export default Nudger;
