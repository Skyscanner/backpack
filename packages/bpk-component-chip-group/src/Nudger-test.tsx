/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { MutableRefObject } from 'react';

import { act, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { CHIP_TYPES } from '../../bpk-component-chip';

import Nudger, { POSITION } from './Nudger';


const mockIsRtl = jest.fn(() => false);

jest.mock('../../bpk-react-utils/index', () => ({
  ...jest.requireActual('../../bpk-react-utils/index'),
  isRTL: () => mockIsRtl(),
}));

const createMockScrollContainerRef = (isRtl: boolean): MutableRefObject<HTMLElement> => ({
  current: {
    scrollBy: jest.fn() as (options?: any) => void,
    offsetWidth: 100,
    scrollLeft: isRtl ? -150 : 150,
    scrollWidth: 500,
  },
} as MutableRefObject<HTMLElement>);

describe('Nudger', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    jest.useFakeTimers();
  });

  it.each([
    [POSITION.trailing, false],
    [POSITION.leading, false],
    [POSITION.trailing, true],
    [POSITION.leading, true],
  ])('should call scrollBy when leading=%s and isRtl=%s', async (position, isRtl) => {
    const user = userEvent.setup({advanceTimers: jest.advanceTimersByTime});
    const mockScrollContainerRef = createMockScrollContainerRef(isRtl);
    mockIsRtl.mockReturnValue(isRtl);
    render(<Nudger ariaLabel="nudge" scrollContainerRef={mockScrollContainerRef} position={position} />);
    await waitFor(() => {
      expect(screen.queryByRole('button')).not.toHaveAttribute('disabled');
    });

    await user.click(screen.getByRole('button'));

    const leading = position === POSITION.leading;
    const isLeft = (leading && !isRtl) || (!leading && isRtl);
    expect(mockScrollContainerRef.current.scrollBy).toHaveBeenCalledTimes(1);
    expect(mockScrollContainerRef.current.scrollBy).toHaveBeenCalledWith({
      left: isLeft ? -150 : 150,
      behavior: 'smooth',
    });
  });

  it('should render button style matching chips',  async () => {
    render(<Nudger position={POSITION.leading} ariaLabel="scroll" scrollContainerRef={createMockScrollContainerRef(false)} chipStyle={CHIP_TYPES.onDark} />);
    await waitFor(() => {
      expect(screen.queryByRole('button')).toBeVisible();
    });

    expect(screen.getByRole('button')).toHaveClass('bpk-button--secondary-on-dark');
  });
});
