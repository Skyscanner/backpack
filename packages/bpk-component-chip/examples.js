/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016-2021 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow strict */

import React from 'react';
import PropTypes from 'prop-types';
import { cssModules } from 'bpk-react-utils';
import BpkText, { TEXT_STYLES } from 'bpk-component-text';
import BeachIconSm from 'bpk-component-icon/sm/beach';
import CarWashIconSm from 'bpk-component-icon/sm/car-wash';
import PlusIconSm from 'bpk-component-icon/sm/plus';
import TickIconSm from 'bpk-component-icon/sm/tick';
import FaceHappyIconSm from 'bpk-component-icon/sm/face--happy';
import { AriaLiveDemo } from 'bpk-component-aria-live/examples';

import STYLES from './examples.scss';
import { COMMON_PROP_TYPES, COMMON_DEFAULT_PROPS } from './src/commonTypes';

import BpkSelectableChip, {
  type BpkSelectableChipProps,
  CHIP_TYPES,
  BpkDismissibleChip,
} from './index';

const getClassName = cssModules(STYLES);

/* ESLint is fighting with Prettier here, so I had to disable it. */
/* eslint-disable flowtype/generic-spacing */
type StatefulSelectableChipProps = $Diff<
  BpkSelectableChipProps,
  {
    accessibilityLabel: string,
    onClick: (event: SyntheticEvent<>) => mixed,
  },
>;
/* eslint-enable */

class StatefulSelectableChip extends React.Component<
  StatefulSelectableChipProps,
  { selected: boolean },
> {
  static propTypes = {
    ...COMMON_PROP_TYPES,
    role: PropTypes.string,
    selected: PropTypes.bool,
    trailingAccessoryView: PropTypes.node,
  };

  static defaultProps = {
    ...COMMON_DEFAULT_PROPS,
    role: 'checkbox',
    selected: false,
    trailingAccessoryView: null,
  };

  constructor(props) {
    super(props);
    this.state = { selected: props.selected };
  }

  toggleSelected = () => {
    this.setState(prevState => {
      return { selected: !prevState.selected };
    });
  };

  render() {
    return (
      <BpkSelectableChip
        {...this.props}
        accessibilityLabel="Toggle chip"
        className={getClassName('bpk-chip-examples__chip')}
        selected={this.state.selected}
        onClick={this.toggleSelected}
      />
    );
  }
}

class StatefulDismissibleChipsExample extends React.Component<
  { type: $Keys<typeof CHIP_TYPES> },
  { chips: Array<string>, updates: Array<string> },
> {
  constructor() {
    super();
    this.state = {
      chips: ['Nara', 'Monteverde', 'Panjin', 'Kolkata'],
      updates: [],
    };
  }

  removeChip = (indexToRemove: number) => {
    this.setState(prevState => {
      const newChips = prevState.chips;
      const removedChip = newChips.splice(indexToRemove, 1)[0];

      return {
        chips: newChips,
        updates: [...prevState.updates, `${removedChip} dismissed.`],
      };
    });
  };

  render() {
    return (
      <div>
        <div>
          {this.state.chips.map((chip, index) => (
            <BpkDismissibleChip
              {...this.props}
              key={index.toString()}
              onClick={() => {
                this.removeChip(index);
              }}
              disabled={index === 3}
              accessibilityLabel="Remove chip"
              className={getClassName('bpk-chip-examples__chip')}
            >
              {chip}
            </BpkDismissibleChip>
          ))}
        </div>
        <AriaLiveDemo
          className={getClassName('bpk-banner-alert-examples__component')}
        >
          {this.state.updates.map(u => (
            <>
              {u}
              <br />
            </>
          ))}
        </AriaLiveDemo>
      </div>
    );
  }
}

const AllSelectableChips = ({ ...rest }) => (
  <div>
    <div className={getClassName('bpk-chip-examples__wrapper')}>
      <StatefulSelectableChip {...rest}>Not selected</StatefulSelectableChip>
      <StatefulSelectableChip leadingAccessoryView={<BeachIconSm />} {...rest}>
        Leading icon
      </StatefulSelectableChip>
      <StatefulSelectableChip
        trailingAccessoryView={<CarWashIconSm />}
        {...rest}
      >
        Trailing icon
      </StatefulSelectableChip>
      <StatefulSelectableChip
        leadingAccessoryView={<BeachIconSm />}
        trailingAccessoryView={<CarWashIconSm />}
        {...rest}
      >
        Both icons
      </StatefulSelectableChip>
      <StatefulSelectableChip selected {...rest}>
        Selected
      </StatefulSelectableChip>
      <StatefulSelectableChip disabled {...rest}>
        Disabled
      </StatefulSelectableChip>
    </div>
  </div>
);

const StatefulSelectableChips = ({ ...rest }: {}) => (
  <div>
    <StatefulSelectableChip {...rest}>Flights</StatefulSelectableChip>
    <StatefulSelectableChip {...rest} selected>
      Hotels
    </StatefulSelectableChip>
    <StatefulSelectableChip {...rest}>Car hire</StatefulSelectableChip>
    <StatefulSelectableChip {...rest} disabled>
      Disabled
    </StatefulSelectableChip>
  </div>
);

const DismissibleChipsExample = () => (
  <StatefulDismissibleChipsExample type={CHIP_TYPES.primary} />
);

const SelectableChipsExample = () => <StatefulSelectableChips />;

const AllSelectableChipStylesExample = () => (
  <div>
    <h3>Primary</h3>
    <StatefulSelectableChips type={CHIP_TYPES.primary} />
    <h3>Success</h3>
    <StatefulSelectableChips type={CHIP_TYPES.success} />
    <h3>Light</h3>
    <StatefulSelectableChips type={CHIP_TYPES.light} />
  </div>
);

const WithIconsExample = () => (
  <div>
    <div>
      <h3>Selectable chips</h3>
      <StatefulSelectableChips
        type={CHIP_TYPES.primary}
        leadingAccessoryView={<PlusIconSm />}
      />
      <StatefulSelectableChips
        type={CHIP_TYPES.primary}
        trailingAccessoryView={<TickIconSm />}
      />
      <StatefulSelectableChips
        type={CHIP_TYPES.primary}
        leadingAccessoryView={<FaceHappyIconSm />}
        trailingAccessoryView={<TickIconSm />}
      />
    </div>
    <div>
      <h3>Dismissable chips</h3>
      <StatefulDismissibleChipsExample
        type={CHIP_TYPES.primary}
        leadingAccessoryView={<FaceHappyIconSm />}
      />
    </div>
  </div>
);

const AllTypesExample = () => (
  <div>
    <BpkText textStyle={TEXT_STYLES.lg} tagName="h2">
      Selectable chips
    </BpkText>
    {Object.keys(CHIP_TYPES).map(chipType => (
      <>
        <BpkText>{chipType}</BpkText>
        <AllSelectableChips type={chipType} />
      </>
    ))}
    <BpkText textStyle={TEXT_STYLES.lg} tagName="h2">
      Dismissible chips
    </BpkText>
    {Object.keys(CHIP_TYPES).map(chipType => (
      <>
        <BpkText>{chipType}</BpkText>
        <StatefulDismissibleChipsExample type={chipType} />
      </>
    ))}
  </div>
);

export {
  AllTypesExample,
  SelectableChipsExample,
  AllSelectableChipStylesExample,
  WithIconsExample,
  DismissibleChipsExample,
};
