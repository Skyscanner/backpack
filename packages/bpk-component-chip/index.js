/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow strict */

import BpkSelectableChip, {
  type Props as SelectableProps,
} from './src/BpkSelectableChip';
import BpkDismissibleChip, {
  type Props as DismissibleProps,
} from './src/BpkDismissibleChip';
import BpkDropdownChip, {
  type Props as DropdownProps,
} from './src/BpkDropdownChip';
import { CHIP_TYPES } from './src/commonTypes';
import themeAttributes from './src/themeAttributes';

export type BpkSelectableChipProps = SelectableProps;
export type BpkDismissibleChipProps = DismissibleProps;
export type BpkDropdownChipProps = DropdownProps;

export default BpkSelectableChip;
export { BpkDismissibleChip, BpkDropdownChip, CHIP_TYPES, themeAttributes };
