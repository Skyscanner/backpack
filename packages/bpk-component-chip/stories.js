/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016-2020 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow strict */

import React from 'react';
import PropTypes from 'prop-types';
import { storiesOf } from '@storybook/react';
import { cssModules } from 'bpk-react-utils';
import BpkText, { TEXT_STYLES } from 'bpk-component-text';
import BeachIconSm from 'bpk-component-icon/sm/beach';
import CarWashIconSm from 'bpk-component-icon/sm/car-wash';

import STYLES from './stories.scss';
import { COMMON_PROP_TYPES, COMMON_DEFAULT_PROPS } from './src/commonTypes';

import BpkSelectableChip, {
  type BpkSelectableChipProps,
  CHIP_TYPES,
  BpkDismissibleChip,
} from './index';

const getClassName = cssModules(STYLES);

/* ESLint is fighting with Prettier here, so I had to disable it. */
/* eslint-disable flowtype/generic-spacing */
type StatefulSelectableChipProps = $Diff<
  BpkSelectableChipProps,
  {
    accessibilityLabel: string,
    onClick: (event: SyntheticEvent<>) => mixed,
  },
>;
/* eslint-enable */

class StatefulSelectableChip extends React.Component<
  StatefulSelectableChipProps,
  { selected: boolean },
> {
  static propTypes = {
    ...COMMON_PROP_TYPES,
    selected: PropTypes.bool,
    trailingAccessoryView: PropTypes.node,
  };

  static defaultProps = {
    ...COMMON_DEFAULT_PROPS,
    selected: false,
    trailingAccessoryView: null,
  };

  constructor(props) {
    super(props);
    this.state = { selected: props.selected };
  }

  toggleSelected = () => {
    this.setState(prevState => {
      return { selected: !prevState.selected };
    });
  };

  render() {
    return (
      <BpkSelectableChip
        {...this.props}
        accessibilityLabel="Toggle chip"
        className={getClassName('bpk-chip-stories__chip')}
        selected={this.state.selected}
        onClick={this.toggleSelected}
      />
    );
  }
}

class StatefulDismissibleChipsExample extends React.Component<
  { type: $Keys<typeof CHIP_TYPES> },
  { chips: Array<string> },
> {
  constructor() {
    super();
    this.state = { chips: ['One', 'Two', 'Three', 'Four'] };
  }

  removeChip = indexToRemove => {
    this.setState(prevState => {
      // I can't figure out how to fix this error, and as it is a non-consumer facing
      // component I don't think it's worth investing time to fix it.
      // $FlowFixMe[incompatible-shape]
      return prevState.chips.splice(indexToRemove, 1);
    });
  };

  render() {
    return (
      <div>
        {this.state.chips.map((chip, index) => (
          <BpkDismissibleChip
            {...this.props}
            key={index.toString()}
            onClick={() => {
              this.removeChip(index);
            }}
            disabled={index === 3}
            accessibilityLabel="Remove chip"
            className={getClassName('bpk-chip-stories__chip')}
          >
            Chip {chip}
          </BpkDismissibleChip>
        ))}
      </div>
    );
  }
}

const ChipsExample = ({ ...rest }) => (
  <div>
    <div className={getClassName('bpk-chip-stories__wrapper')}>
      <StatefulSelectableChip {...rest}>Not selected</StatefulSelectableChip>
      <StatefulSelectableChip leadingAccessoryView={<BeachIconSm />} {...rest}>
        Leading icon
      </StatefulSelectableChip>
      <StatefulSelectableChip
        trailingAccessoryView={<CarWashIconSm />}
        {...rest}
      >
        Trailing icon
      </StatefulSelectableChip>
      <StatefulSelectableChip
        leadingAccessoryView={<BeachIconSm />}
        trailingAccessoryView={<CarWashIconSm />}
        {...rest}
      >
        Both icons
      </StatefulSelectableChip>
      <StatefulSelectableChip selected {...rest}>
        Selected
      </StatefulSelectableChip>
      <StatefulSelectableChip disabled {...rest}>
        Disabled
      </StatefulSelectableChip>
    </div>
  </div>
);

storiesOf('bpk-component-chip', module).add('Default', () => (
  <div>
    <BpkText textStyle={TEXT_STYLES.lg} tagName="h2">
      Selectable chips
    </BpkText>
    {Object.keys(CHIP_TYPES).map(chipType => (
      <>
        <BpkText>{chipType}</BpkText>
        <ChipsExample type={chipType} />
      </>
    ))}
    <BpkText textStyle={TEXT_STYLES.lg} tagName="h2">
      Dismissible chips
    </BpkText>
    {Object.keys(CHIP_TYPES).map(chipType => (
      <>
        <BpkText>{chipType}</BpkText>
        <StatefulDismissibleChipsExample type={chipType} />
      </>
    ))}
  </div>
));
