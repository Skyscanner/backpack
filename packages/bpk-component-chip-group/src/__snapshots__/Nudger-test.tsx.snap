// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nudger should render correctly for onDark chip style 1`] = `
<DocumentFragment>
  <div
    class="bpk-chip-group-nudger bpk-chip-group-nudger--trailing bpk-chip-group-nudger--hidden"
  >
    <button
      class="bpk-button bpk-button--icon-only bpk-button--secondary-on-dark"
      disabled=""
      type="button"
    >
      <span
        style="line-height: 1rem; display: inline-block; margin-top: 0.25rem; vertical-align: top;"
      >
        <svg
          aria-hidden="true"
          style="width: 1rem; height: 1rem;"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M3 12a1.5 1.5 0 001.5 1.5h11.379l-4.94 4.94a1.5 1.5 0 002.122 2.12l7.5-7.5a1.5 1.5 0 000-2.12l-7.5-7.5a1.5 1.5 0 00-2.122 2.12l4.94 4.94H4.5A1.5 1.5 0 003 12z"
          />
        </svg>
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Nudger should render correctly when leading=false and isRtl=false 1`] = `
<DocumentFragment>
  <div
    class="bpk-chip-group-nudger bpk-chip-group-nudger--trailing bpk-chip-group-nudger--hidden"
  >
    <button
      class="bpk-button bpk-button--icon-only bpk-button--secondary"
      disabled=""
      type="button"
    >
      <span
        style="line-height: 1rem; display: inline-block; margin-top: 0.25rem; vertical-align: top;"
      >
        <svg
          aria-hidden="true"
          style="width: 1rem; height: 1rem;"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M3 12a1.5 1.5 0 001.5 1.5h11.379l-4.94 4.94a1.5 1.5 0 002.122 2.12l7.5-7.5a1.5 1.5 0 000-2.12l-7.5-7.5a1.5 1.5 0 00-2.122 2.12l4.94 4.94H4.5A1.5 1.5 0 003 12z"
          />
        </svg>
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Nudger should render correctly when leading=false and isRtl=true 1`] = `
<DocumentFragment>
  <div
    class="bpk-chip-group-nudger bpk-chip-group-nudger--trailing bpk-chip-group-nudger--hidden"
  >
    <button
      class="bpk-button bpk-button--icon-only bpk-button--secondary"
      disabled=""
      type="button"
    >
      <span
        style="line-height: 1rem; display: inline-block; margin-top: 0.25rem; vertical-align: top;"
      >
        <svg
          aria-hidden="true"
          style="width: 1rem; height: 1rem;"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M21 12a1.5 1.5 0 01-1.5 1.5H8.121l4.94 4.94a1.5 1.5 0 01-2.122 2.12l-7.5-7.5a1.5 1.5 0 010-2.12l7.5-7.5a1.5 1.5 0 012.122 2.12L8.12 10.5H19.5A1.5 1.5 0 0121 12z"
          />
        </svg>
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Nudger should render correctly when leading=true and isRtl=false 1`] = `
<DocumentFragment>
  <div
    class="bpk-chip-group-nudger bpk-chip-group-nudger--leading bpk-chip-group-nudger--hidden"
  >
    <button
      class="bpk-button bpk-button--icon-only bpk-button--secondary"
      disabled=""
      type="button"
    >
      <span
        style="line-height: 1rem; display: inline-block; margin-top: 0.25rem; vertical-align: top;"
      >
        <svg
          aria-hidden="true"
          style="width: 1rem; height: 1rem;"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M21 12a1.5 1.5 0 01-1.5 1.5H8.121l4.94 4.94a1.5 1.5 0 01-2.122 2.12l-7.5-7.5a1.5 1.5 0 010-2.12l7.5-7.5a1.5 1.5 0 012.122 2.12L8.12 10.5H19.5A1.5 1.5 0 0121 12z"
          />
        </svg>
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Nudger should render correctly when leading=true and isRtl=true 1`] = `
<DocumentFragment>
  <div
    class="bpk-chip-group-nudger bpk-chip-group-nudger--leading bpk-chip-group-nudger--hidden"
  >
    <button
      class="bpk-button bpk-button--icon-only bpk-button--secondary"
      disabled=""
      type="button"
    >
      <span
        style="line-height: 1rem; display: inline-block; margin-top: 0.25rem; vertical-align: top;"
      >
        <svg
          aria-hidden="true"
          style="width: 1rem; height: 1rem;"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M3 12a1.5 1.5 0 001.5 1.5h11.379l-4.94 4.94a1.5 1.5 0 002.122 2.12l7.5-7.5a1.5 1.5 0 000-2.12l-7.5-7.5a1.5 1.5 0 00-2.122 2.12l4.94 4.94H4.5A1.5 1.5 0 003 12z"
          />
        </svg>
      </span>
    </button>
  </div>
</DocumentFragment>
`;
